### PR TITLE
Convert `Delegator` base class to JS

### PR DESCRIPTION
**Pull request by robertknight**
Convert the `Delegator` base class to JS and rewrite the documentation
to match current functionality and usage. Also simplify the `options`
initialization and add a note about why the constructor reads
`this.options` before assigning it.

The event pub/sub functionality doesn't really need jQuery any more, but changing that can happen separately.

I tried to convert the `Plugin` class as well but encountered an issue
with interaction between Babel-compiled derived classes and CoffeeScript
sub-classes (in particular, the `BucketBar` class), so it looks like
`Plugin` sub-classes will need to be converted first.

### CODE REVIEW

File: src/annotator/delegator.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+//  Copyright 2015, the Annotator project contributors.
+//  Dual licensed under the MIT and GPLv3 licenses.
+//  https://github.com/openannotation/annotator/blob/master/LICENSE
+
+/**
+ * `Delegator` is a base class for many objects in the annotator.
+ *
+ * It provides:
+ *
+ *  - An event bus, attached to the DOM element passed to the constructor.
```
Comment (lyzadanger) [inline src/annotator/delegator.js:18]: In practice across annotator classes, is this element always the same root element? Curious.

File: src/annotator/delegator.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+export default class Delegator {
+  /**
+   * Construct the `Delegator` instance.
+   *
+   * @param {HTMLElement} element
+   * @param {Object} [config]
+   */
+  constructor(element, config) {
+    // Some subclasses rely on a legacy mechanism of defining options at
+    // the class level by defining an `options` property on the prototype.
```
Comment (lyzadanger) [inline src/annotator/delegator.js:35]: Will this eventually go away when we migrate/refactor subclasses? (p.s. thanks for this comment; I could see this raising the questioning eyebrows of future devs)

File: src/annotator/delegator.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const classOptions = this.options || {};
+    this.options = { ...classOptions, ...config };
+    this.element = $(element);
```
Comment (lyzadanger) [inline src/annotator/delegator.js:38]: hmmm. I suppose the naming of `this.element` is...likely baked into subclasses? That this is quietly a jQuery element gives me the clenches, but...with luck, jQuery is not long for our world...

File: src/annotator/delegator.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  /**
+   * Fire an event.
+   *
+   * This and other `Delegator` instances which share the same root element will
+   * be able to observe it.
+   *
+   * @param {string} event
+   * @param {any[]} args
```
Comment (lyzadanger) [inline src/annotator/delegator.js:60]: Should `args` be optional here?

File: src/annotator/delegator.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    // Wrapper that strips the `event` argument.
+    const closure = (event, ...args) => callback(...args);
+
+    // Ensure both functions have the same unique id so that jQuery will accept
+    // callback when unbinding closure.
+    //
+    // @ts-expect-error - `guid` property is non-standard
+    closure.guid = callback.guid = $.guid += 1;
+
+    // eslint-disable-next-line
```
Comment (lyzadanger) [inline src/annotator/delegator.js:82]: Why?

**Review by lyzadanger — APPROVED** [review]
Looking good; just a few questions/comments sprinkled in for zest.

File: src/annotator/delegator.js (hunk 1 above)
Comment (robertknight) [reply src/annotator/delegator.js:18]: All the classes which need to publish or subscribe to events will be sharing the same root element, I haven't checked if those classes which don't deal with events at all are.

File: src/annotator/delegator.js (hunk 2 above)
Comment (robertknight) [reply src/annotator/delegator.js:35]: Yes. To begin with I expect we'll require subclasses to pass all options in the `config` parameter.

File: src/annotator/delegator.js (hunk 3 above)
Comment (robertknight) [reply src/annotator/delegator.js:38]: Yes, `this.element` / `@element` is a public API that is used by subclasses.

File: src/annotator/delegator.js (hunk 4 above)
Comment (robertknight) [reply src/annotator/delegator.js:60]: Indeed!

File: src/annotator/delegator.js (hunk 5 above)
Comment (robertknight) [reply src/annotator/delegator.js:82]: I'll add a note - the next line produces a false positive for use of the `bind` method of functions, whereas this is a different function.

**Pull request merged**